lib/wallet/src/Cardano/Wallet/DB/Layer.hs

```diff
@@ -757,70 +759,90 @@ readWalletMetadata wid =
-                (StakeKeyCertificateKey wid sl)
-                (StakeKeyCertificate wid sl W.StakeKeyRegistration)
-
-    , putDelegationRewardBalance_ = \amt ->
+
+mkDBDelegation ::
+    TimeInterpreter IO -> W.WalletId -> DBDelegation (SqlPersistT IO)
+mkDBDelegation ti wid =
+    DBDelegation
+        { isStakeKeyRegistered_
+        , putDelegationCertificate_
+        , putDelegationRewardBalance_
+        , readDelegationRewardBalance_
+        , readDelegation_
+        }
+  where
+    isStakeKeyRegistered_ :: SqlPersistT IO Bool
+    isStakeKeyRegistered_ = do
+        val <- fmap entityVal <$>
+            selectFirst [StakeKeyCertWalletId ==. wid] [Desc StakeKeyCertSlot]
+        pure $
+            case val of
+                Nothing -> False
+                Just (StakeKeyCertificate _ _ status) ->
+                    status == W.StakeKeyRegistration
+
+    putDelegationCertificate_ ::
+        W.DelegationCertificate -> W.SlotNo -> SqlPersistT IO ()
+    putDelegationCertificate_ cert sl =
+        case cert of
+            W.CertDelegateNone _ -> do
+                repsert
+                    (DelegationCertificateKey wid sl)
+                    (DelegationCertificate wid sl Nothing)
+                repsert
+                    (StakeKeyCertificateKey wid sl)
+                    (StakeKeyCertificate wid sl W.StakeKeyDeregistration)
+            W.CertDelegateFull _ pool ->
+                repsert
+                    (DelegationCertificateKey wid sl)
+                    (DelegationCertificate wid sl (Just pool))
+            W.CertRegisterKey _ ->
+                repsert
+                    (StakeKeyCertificateKey wid sl)
+                    (StakeKeyCertificate wid sl W.StakeKeyRegistration)
+
+    putDelegationRewardBalance_ :: Coin.Coin -> SqlPersistT IO ()
+    putDelegationRewardBalance_ amount =
         repsert
             (DelegationRewardKey wid)
-            (DelegationReward wid (Coin.unsafeToWord64 amt))
+            (DelegationReward wid (Coin.unsafeToWord64 amount))
 
-    , readDelegationRewardBalance_ =
+    readDelegationRewardBalance_ :: SqlPersistT IO Coin.Coin
+    readDelegationRewardBalance_ =
         Coin.fromWord64 . maybe 0 (rewardAccountBalance . entityVal) <$>
             selectFirst [RewardWalletId ==. wid] []
-    , readDelegation_ = \epoch -> if
-        | epoch == 0 -> pure $ W.WalletDelegation W.NotDelegating []
-        | otherwise -> do
-            (eMinus1, e) <- liftIO $ interpretQuery ti $
-                (,) <$> firstSlotInEpoch (epoch - 1) <*> firstSlotInEpoch epoch
-            active <- maybe W.NotDelegating toWalletDelegationStatus
-                <$> readDelegationCertificate wid
-                    [ CertSlot <. eMinus1
+
+    readDelegation_ :: W.EpochNo -> SqlPersistT IO W.WalletDelegation
+    readDelegation_ epoch = case epoch of
+        0 -> do
+            currEpochStartSlot <-
+                liftIO $ interpretQuery ti $ firstSlotInEpoch epoch
```

Comment:
Might be worth factoring out this and the corresponding query in the other branch to outside the case statement.

Reply:
Actually, branches have a slightly different `Qry`s: The branch where the `epoch != 0` interprets the `Qry (SlotNo, SlotNo)` whereas the branch `epoch == 0` interprets `Qry SlotNo`. I can see how its possible to always query for both data points and not use `prevEpochStartSlot` when `epoch == 0`, but I don't feel confident to say that such over-fetching has zero cost.

Reply:
Ah right, I suspect `prevEpochStartSlot when epoch == 0` will fail for `epoch == 0`, so it might not make sense to factor out then

Reply:
Indeed, it fails with the `TimeHorizonException` or alike.